## 4.8 Group Data & Aggregate Variables

### This script contains the following points:

#### 1. Import data & Check dimensions - create subset for practice
#### 2. Grouping data with pandas
#### 3. Aggregating data with agg()
#### 4. Create Customer Loyalty flag/label columns
#### 5. Task Questions
#### 6. Export file

### Importing libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

### 01. & 02. Importing data & Checking dimensions

In [2]:
# Create string path for main project folder

path = r'/Users/AngieUS/Desktop/Instacart Project'

# Import ords_prods_merge data (ords_prods_merge_nv.pkl data file from last task)

df = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_nv.pkl'))

# Check df

df.head()  

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2398795       1             2                   3                  7   
1  2398795       1             2                   3                  7   
2  2398795       1             2                   3                  7   
3  2398795       1             2                   3                  7   
4  2398795       1             2                   3                  7   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                    15.0        196                  1          1   
1                    15.0      10258                  2          0   
2                    15.0      12427                  3          1   
3                    15.0      13176                  4          0   
4                    15.0      26088                  5          1   

                 product_name aisle_id department_id  prices _merge  \
0                        Soda       77             7     9.0   both   
1                  Pistachios      117            19     3.0   both   
2         Original Beef Jerky       23            19     4.4   both   
3      Bag of Organic Bananas       24             4    10.3   both   
4  Aged White Cheddar Popcorn       23            19     4.7   both   

         price_range busiest_days busiest_period_of_day  
0  Mid-range product   Least busy        Average orders  
1  Low-range product   Least busy        Average orders  
2  Low-range product   Least busy        Average orders  
3  Mid-range product   Least busy        Average orders  
4  Low-range product   Least busy        Average orders

In [3]:
# Display dimensions of df - rows, columns

df.shape

(30328763, 17)

In [4]:
# Create a subset of df

df_sub = df[:1000000]

# Display dimensions of df_sub - rows, columns

df_sub.shape   

(1000000, 17)

### 03. Grouping data with pandas

In [5]:
# Step 1: Group a dataframe by a column

df_sub.groupby('product_name')

### 04. Aggregating data with agg()

In [6]:
# Step 1: group by department id
# Step 2: Apply a function or aggregate the data
# Performing a single aggregation

df_sub.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.937254
10               21.159963
11               17.562788
12               16.716021
13               17.624550
14               18.692641
15               17.081722
16               19.059910
17               17.411197
18               20.550802
19               18.810610
2                18.365385
20               18.384576
21               22.424189
3                19.176605
4                18.967580
5                16.163447
6                16.547793
7                18.897694
8                17.262475
9                17.148459

In [9]:
# Using the mean() function to also aggregate by dept id

df_sub.groupby('department_id')['order_number'].mean()

department_id
1     15.937254
10    21.159963
11    17.562788
12    16.716021
13    17.624550
14    18.692641
15    17.081722
16    19.059910
17    17.411197
18    20.550802
19    18.810610
2     18.365385
20    18.384576
21    22.424189
3     19.176605
4     18.967580
5     16.163447
6     16.547793
7     18.897694
8     17.262475
9     17.148459
Name: order_number, dtype: float64

In [7]:
# Performing multiple aggregations (mean, min, max)

df_sub.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.937254   2  99
10               21.159963   2  99
11               17.562788   2  99
12               16.716021   2  99
13               17.624550   2  99
14               18.692641   2  99
15               17.081722   2  99
16               19.059910   2  99
17               17.411197   2  99
18               20.550802   2  99
19               18.810610   2  99
2                18.365385   2  98
20               18.384576   2  99
21               22.424189   2  97
3                19.176605   2  99
4                18.967580   2  99
5                16.163447   2  99
6                16.547793   2  99
7                18.897694   2  99
8                17.262475   2  91
9                17.148459   2  99

### 05. Create Customer Loyalty flag/label columns

In [8]:
# Aggregating data with transform() & loc() - "Loyalty Customers"
# Creating a flag for customers with a lot of orders = loyal customers = max orders > 40

# Step 1 - split the data into groups based on user_id column
# Step 2 - use transform() function on the order_number column
# Step 3 - create new column (max_order) to place results of the aggregation
# - written in one line of code. note: df = ords_prods_merge

df['max_order'] = df.groupby(['user_id'])['order_number'].transform('max')

In [9]:
# Check values

df['max_order'].value_counts(dropna=False)

max_order
99    1157474
9      720052
8      710465
11     699701
10     696026
       ...   
90      47748
97      44452
98      44112
96      40045
2           4
Name: count, Length: 98, dtype: int64

In [10]:
# Check df

df.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2398795       1             2                   3                  7   
1  2398795       1             2                   3                  7   
2  2398795       1             2                   3                  7   
3  2398795       1             2                   3                  7   
4  2398795       1             2                   3                  7   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                    15.0        196                  1          1   
1                    15.0      10258                  2          0   
2                    15.0      12427                  3          1   
3                    15.0      13176                  4          0   
4                    15.0      26088                  5          1   

                 product_name aisle_id department_id  prices _merge  \
0                        Soda       77             7     9.0   both   
1                  Pistachios      117            19     3.0   both   
2         Original Beef Jerky       23            19     4.4   both   
3      Bag of Organic Bananas       24             4    10.3   both   
4  Aged White Cheddar Popcorn       23            19     4.7   both   

         price_range busiest_days busiest_period_of_day  max_order  
0  Mid-range product   Least busy        Average orders         10  
1  Low-range product   Least busy        Average orders         10  
2  Low-range product   Least busy        Average orders         10  
3  Mid-range product   Least busy        Average orders         10  
4  Low-range product   Least busy        Average orders         10

In [11]:
# Create new column to assign loyalty label to user_id based on max_order value
# Criteria #1

df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [12]:
# Criteria #2

df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [13]:
# Criteria #3

df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [14]:
# Check values
df['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular customer    15081691
Loyal customer      10095381
New customer         5151691
Name: count, dtype: int64

In [15]:
# Check df

df.head ()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2398795       1             2                   3                  7   
1  2398795       1             2                   3                  7   
2  2398795       1             2                   3                  7   
3  2398795       1             2                   3                  7   
4  2398795       1             2                   3                  7   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                    15.0        196                  1          1   
1                    15.0      10258                  2          0   
2                    15.0      12427                  3          1   
3                    15.0      13176                  4          0   
4                    15.0      26088                  5          1   

                 product_name aisle_id department_id  prices _merge  \
0                        Soda       77             7     9.0   both   
1                  Pistachios      117            19     3.0   both   
2         Original Beef Jerky       23            19     4.4   both   
3      Bag of Organic Bananas       24             4    10.3   both   
4  Aged White Cheddar Popcorn       23            19     4.7   both   

         price_range busiest_days busiest_period_of_day  max_order  \
0  Mid-range product   Least busy        Average orders         10   
1  Low-range product   Least busy        Average orders         10   
2  Low-range product   Least busy        Average orders         10   
3  Mid-range product   Least busy        Average orders         10   
4  Low-range product   Least busy        Average orders         10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer

### 05. Task Questions

In [16]:
# Display aggregated mean for entire dataframe (df = ords.prods.merge)
# Step 1: group by department id
# Step 2: Apply a function or aggregate the data
# Performing a single aggregation

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                16.559358
10               21.227447
11               17.311768
12               16.953613
13               17.727470
14               17.898097
15               17.290123
16               18.757796
17               16.808683
18               20.336389
19               18.258620
2                18.413176
20               17.578959
21               23.644477
3                18.279600
4                18.915890
5                16.497751
6                17.609390
7                18.303975
8                16.383301
9                17.022963

### The mean of the subset in comparison to the entire data set is very close - within (<) 1 order. There is only one department (6) that varies slightly over 1 at 1.06.

In [17]:
# Display basic pricing statistics for the different loyalty categories to gain insights on how they spend

df.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.402162  1.0  99999.0
New customer      13.396333  1.0  99999.0
Regular customer  12.546842  1.0  99999.0

In [18]:
df.groupby('loyalty_flag').describe()

order_number                                               \
                        count       mean        std  min   25%   50%   75%   
loyalty_flag                                                                 
Loyal customer     10095381.0  33.232662  21.461954  2.0  16.0  31.0  46.0   
New customer        5151691.0   4.402213   2.088623  2.0   3.0   4.0   6.0   
Regular customer   15081691.0  12.945673   8.284392  2.0   6.0  11.0  18.0   

                       orders_day_of_week            ... prices           \
                   max              count      mean  ...    75%      max   
loyalty_flag                                         ...                   
Loyal customer    99.0         10095381.0  2.791594  ...   11.2  99999.0   
New customer      10.0          5151691.0  2.696423  ...   11.3  99999.0   
Regular customer  40.0         15081691.0  2.722023  ...   11.3  99999.0   

                   max_order                                                \
                       count       mean        std   min   25%   50%   75%   
loyalty_flag                                                                 
Loyal customer    10095381.0  64.063418  18.815551  41.0  48.0  58.0  77.0   
New customer       5151691.0   6.758699   2.212342   2.0   5.0   7.0   9.0   
Regular customer  15081691.0  23.587759   8.528039  11.0  16.0  23.0  31.0   

                        
                   max  
loyalty_flag            
Loyal customer    99.0  
New customer      10.0  
Regular customer  40.0  

[3 rows x 64 columns]

### Reviewing the aggregated pricing data for the 3 different loyalty groups - there is a difference in their spending as evidenced by the mean for each. Make note: The current tasks and lessons have not given instruction or requirement to clean records of obvious issue. Specifically, there are 4120 records with a price of 14900 and 656 records with a price of 99999. While it's highly doubtful there is any item at instacart at these prices - technically I have no means of research to delete these.

In [19]:
# "Customer Spending Tier"
# Creating a flag for customers who spend a lot = prices > 10

# Step 1 - split the data into groups based on user_id column
# Step 2 - use transform() function on the prices column
# Step 3 - create new column (avg_prices) to place results of the aggregation
# - written in one line of code. note: df = ords_prods_merge

df['avg_prices'] = df.groupby(['user_id'])['prices'].transform('mean')

In [20]:
# Check values

df['avg_prices'].value_counts(dropna=False)

avg_prices
8.00     9353
8.40     7568
7.80     7241
7.50     7226
8.10     6550
         ... 
14.20       2
12.90       2
14.15       2
1.95        2
2.75        2
Name: count, Length: 121423, dtype: int64

In [21]:
# Check df

df.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2398795       1             2                   3                  7   
1  2398795       1             2                   3                  7   
2  2398795       1             2                   3                  7   
3  2398795       1             2                   3                  7   
4  2398795       1             2                   3                  7   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                    15.0        196                  1          1   
1                    15.0      10258                  2          0   
2                    15.0      12427                  3          1   
3                    15.0      13176                  4          0   
4                    15.0      26088                  5          1   

                 product_name aisle_id department_id  prices _merge  \
0                        Soda       77             7     9.0   both   
1                  Pistachios      117            19     3.0   both   
2         Original Beef Jerky       23            19     4.4   both   
3      Bag of Organic Bananas       24             4    10.3   both   
4  Aged White Cheddar Popcorn       23            19     4.7   both   

         price_range busiest_days busiest_period_of_day  max_order  \
0  Mid-range product   Least busy        Average orders         10   
1  Low-range product   Least busy        Average orders         10   
2  Low-range product   Least busy        Average orders         10   
3  Mid-range product   Least busy        Average orders         10   
4  Low-range product   Least busy        Average orders         10   

   loyalty_flag  avg_prices  
0  New customer    6.372222  
1  New customer    6.372222  
2  New customer    6.372222  
3  New customer    6.372222  
4  New customer    6.372222

In [22]:
# Create new column to assign spending label to user_id based on max_prices value
# Criteria #1

df.loc[df['avg_prices'] < 10, 'spending_flag'] = 'Low Spender'

In [23]:
# Criteria #2

df.loc[df['avg_prices'] >= 10, 'spending_flag'] = 'High Spender'

In [24]:
# Check values

df['spending_flag'].value_counts(dropna=False)

spending_flag
Low Spender     29730505
High Spender      598258
Name: count, dtype: int64

In [25]:
# Check df

df.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2398795       1             2                   3                  7   
1  2398795       1             2                   3                  7   
2  2398795       1             2                   3                  7   
3  2398795       1             2                   3                  7   
4  2398795       1             2                   3                  7   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                    15.0        196                  1          1   
1                    15.0      10258                  2          0   
2                    15.0      12427                  3          1   
3                    15.0      13176                  4          0   
4                    15.0      26088                  5          1   

                 product_name  ... department_id prices  _merge  \
0                        Soda  ...             7    9.0    both   
1                  Pistachios  ...            19    3.0    both   
2         Original Beef Jerky  ...            19    4.4    both   
3      Bag of Organic Bananas  ...             4   10.3    both   
4  Aged White Cheddar Popcorn  ...            19    4.7    both   

         price_range busiest_days busiest_period_of_day max_order  \
0  Mid-range product   Least busy        Average orders        10   
1  Low-range product   Least busy        Average orders        10   
2  Low-range product   Least busy        Average orders        10   
3  Mid-range product   Least busy        Average orders        10   
4  Low-range product   Least busy        Average orders        10   

   loyalty_flag avg_prices  spending_flag  
0  New customer   6.372222    Low Spender  
1  New customer   6.372222    Low Spender  
2  New customer   6.372222    Low Spender  
3  New customer   6.372222    Low Spender  
4  New customer   6.372222    Low Spender  

[5 rows x 21 columns]

In [26]:
# "Customer Frequency Tier"
# Creating a flag for customers who order regularly = days since last order <= 10

# Step 1 - split the data into groups based on user_id column
# Step 2 - use transform() function on the days_since_prior_order column
# Step 3 - create new column (median_days) to place results of the aggregation
# - written in one line of code. note: df = ords_prods_merge

df['median_days'] = df.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [27]:
# Check values

df['median_days'].value_counts(dropna=False)

median_days
7.0     5445701
6.0     2870088
8.0     2571811
5.0     2405138
4.0     2252267
         ...   
27.5       4668
26.5       4534
29.5       4196
1.5        1556
0.5         178
Name: count, Length: 61, dtype: int64

In [28]:
# Check df

df.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2398795       1             2                   3                  7   
1  2398795       1             2                   3                  7   
2  2398795       1             2                   3                  7   
3  2398795       1             2                   3                  7   
4  2398795       1             2                   3                  7   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                    15.0        196                  1          1   
1                    15.0      10258                  2          0   
2                    15.0      12427                  3          1   
3                    15.0      13176                  4          0   
4                    15.0      26088                  5          1   

                 product_name  ... prices _merge        price_range  \
0                        Soda  ...    9.0   both  Mid-range product   
1                  Pistachios  ...    3.0   both  Low-range product   
2         Original Beef Jerky  ...    4.4   both  Low-range product   
3      Bag of Organic Bananas  ...   10.3   both  Mid-range product   
4  Aged White Cheddar Popcorn  ...    4.7   both  Low-range product   

  busiest_days busiest_period_of_day max_order  loyalty_flag  avg_prices  \
0   Least busy        Average orders        10  New customer    6.372222   
1   Least busy        Average orders        10  New customer    6.372222   
2   Least busy        Average orders        10  New customer    6.372222   
3   Least busy        Average orders        10  New customer    6.372222   
4   Least busy        Average orders        10  New customer    6.372222   

  spending_flag  median_days  
0   Low Spender         20.5  
1   Low Spender         20.5  
2   Low Spender         20.5  
3   Low Spender         20.5  
4   Low Spender         20.5  

[5 rows x 22 columns]

In [29]:
# Create new column to assign order_freq_flag label to user_id based on median_days value
# Criteria #1

df.loc[df['median_days'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [30]:
# Criteria #2

df.loc[(df['median_days'] <= 20) & (df['median_days'] > 10), 'order_freq_flag'] = 'Regular customer'

In [33]:
# Criteria #3

df.loc[df['median_days'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [34]:
# Check values

df['order_freq_flag'].value_counts(dropna=False)

order_freq_flag
Frequent customer        20675290
Regular customer          6594542
Non-frequent customer     3058931
Name: count, dtype: int64

In [35]:
# Check df

df.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2398795       1             2                   3                  7   
1  2398795       1             2                   3                  7   
2  2398795       1             2                   3                  7   
3  2398795       1             2                   3                  7   
4  2398795       1             2                   3                  7   

   days_since_prior_order product_id  add_to_cart_order  reordered  \
0                    15.0        196                  1          1   
1                    15.0      10258                  2          0   
2                    15.0      12427                  3          1   
3                    15.0      13176                  4          0   
4                    15.0      26088                  5          1   

                 product_name  ... _merge        price_range  busiest_days  \
0                        Soda  ...   both  Mid-range product    Least busy   
1                  Pistachios  ...   both  Low-range product    Least busy   
2         Original Beef Jerky  ...   both  Low-range product    Least busy   
3      Bag of Organic Bananas  ...   both  Mid-range product    Least busy   
4  Aged White Cheddar Popcorn  ...   both  Low-range product    Least busy   

  busiest_period_of_day max_order  loyalty_flag avg_prices  spending_flag  \
0        Average orders        10  New customer   6.372222    Low Spender   
1        Average orders        10  New customer   6.372222    Low Spender   
2        Average orders        10  New customer   6.372222    Low Spender   
3        Average orders        10  New customer   6.372222    Low Spender   
4        Average orders        10  New customer   6.372222    Low Spender   

  median_days        order_freq_flag  
0        20.5  Non-frequent customer  
1        20.5  Non-frequent customer  
2        20.5  Non-frequent customer  
3        20.5  Non-frequent customer  
4        20.5  Non-frequent customer  

[5 rows x 23 columns]

In [36]:
# Display dimensions of df - rows, columns

df.shape

(30328763, 23)

### 06. Export file

In [37]:
# Export data (pkl file) as ords_prods_merge_agg

df.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_agg.pkl'))